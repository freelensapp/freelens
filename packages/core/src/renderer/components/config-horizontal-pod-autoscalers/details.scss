/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "autoscaler.mixins";
@use "../vars" as *;

.HpaDetails {
  .status {
    @include autoscaler.hpa-status-bgc;
  }

  .metrics .Table {
    margin: 0 (-$margin * 3);

    .TableCell {
      word-break: break-word;


      &:first-child {
        margin-left: $margin * 2;
      }

      &:last-child {
        margin-right: $margin * 2;
      }

      &.name {
        flex-grow: 2;
      }
    }
  }
}
