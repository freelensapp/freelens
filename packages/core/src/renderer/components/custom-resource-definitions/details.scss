/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "mixins.scss";
@use "../vars" as *;

.CustomResourceDefinitionDetails {
  .conditions {
    @include mixins.custom-resource-definitions-condition-colors;
  }

  .Table {
    margin-left: -$margin * 2;
    margin-right: -$margin * 2;

    .TableRow {
      cursor: default;
    }

    &.printer-columns {
      .type {
        flex: 0.5;
      }

      .json-path {
        flex: 3;
      }
    }
  }
}
