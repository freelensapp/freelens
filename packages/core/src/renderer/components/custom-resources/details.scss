/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

$custom-resource-condition-colors: (
  Accepted: var(--colorOk),
  Available: var(--colorOk),
  Ready: var(--colorSuccess),
  ReconcileSuccess: var(--colorOk),
  Succeeded: var(--colorOk),
  Success: var(--colorOk),
  UpgradeSucceeded: var(--colorOk),
);

.CustomResourceDetails {
  .conditions {
    @each $status, $color in $custom-resource-condition-colors {
      .#{$status} {
        background: $color;
        color: white;
      }
    }
  }
}
