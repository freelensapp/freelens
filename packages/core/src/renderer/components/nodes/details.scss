/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

@use "nodes-mixins";

.NodeDetails {
  --status-ready-bg: var(--colorOk);
  --status-default-bg: var(--colorError);

  .conditions {
    @include nodes-mixins.node-status-bgs;
  }
}
