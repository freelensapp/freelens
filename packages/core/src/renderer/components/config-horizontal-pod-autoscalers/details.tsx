/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import type { HorizontalPodAutoscalerMetricSpec, HorizontalPodAutoscalerMetricTarget } from "@freelensapp/kube-object";
import { HorizontalPodAutoscaler } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import { cssNames } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Link } from "react-router-dom";
import type { ApiManager } from "../../../common/k8s-api/api-manager";
import apiManagerInjectable from "../../../common/k8s-api/api-manager/manager.injectable";
import { Badge } from "../badge";
import { DrawerItem, DrawerTitle } from "../drawer";
import type { GetDetailsUrl } from "../kube-detail-params/get-details-url.injectable";
import getDetailsUrlInjectable from "../kube-detail-params/get-details-url.injectable";
import type { KubeObjectDetailsProps } from "../kube-object-details";
import { Table, TableCell, TableHead, TableRow } from "../table";
import { getMetricName } from "./get-metric-name";
import getHorizontalPodAutoscalerMetrics from "./get-metrics.injectable";

export interface HpaDetailsProps extends KubeObjectDetailsProps<HorizontalPodAutoscaler> {}

interface Dependencies {
  apiManager: ApiManager;
  logger: Logger;
  getDetailsUrl: GetDetailsUrl;
  getMetrics: (hpa: HorizontalPodAutoscaler) => string[];
}

@observer
class NonInjectedHorizontalPodAutoscalerDetails extends React.Component<HpaDetailsProps & Dependencies> {
  private renderTargetLink(target: HorizontalPodAutoscalerMetricTarget | undefined) {
    if (!target) {
      return null;
    }

    const { object: hpa, apiManager, getDetailsUrl } = this.props;
    const { kind, name } = target;
    const objectUrl = getDetailsUrl(apiManager.lookupApiLink(target, hpa));

    return (
      <>
        on
        <Link to={objectUrl}>{`${kind}/${name}`}</Link>
      </>
    );
  }

  renderMetrics() {
    const { object: hpa } = this.props;

    const renderName = (metric: HorizontalPodAutoscalerMetricSpec) => {
      const metricName = getMetricName(metric);

      switch (metric?.type) {
        case "ContainerResource":

        // fallthrough
        case "Resource": {
          const metricSpec = metric.resource ?? metric.containerResource;

          return `Resource ${metricSpec.name} on Pods`;
        }
        case "Pods":
          return `${metricName ?? ""} on Pods`;

        case "Object": {
          return (
            <>
              {metricName} {this.renderTargetLink(metric.object?.describedObject)}
            </>
          );
        }
        case "External":
          return `${metricName ?? ""} on ${JSON.stringify(metric.external.metricSelector ?? metric.external.metric?.selector)}`;
        default:
          return hpa.spec?.targetCPUUtilizationPercentage ? "CPU Utilization percentage" : "unknown";
      }
    };

    return (
      <Table data-testid="hpa-metrics">
        <TableHead flat>
          <TableCell className="name">Name</TableCell>
          <TableCell className="metrics">Current / Target</TableCell>
        </TableHead>
        {this.props.getMetrics(hpa).map((metrics, index) => (
          <TableRow key={index}>
            <TableCell className="name">{renderName(hpa.getMetrics()[index])}</TableCell>
            <TableCell className="metrics">{metrics}</TableCell>
          </TableRow>
        ))}
      </Table>
    );
  }

  render() {
    const { object: hpa, apiManager, getDetailsUrl, logger } = this.props;

    if (!hpa) {
      return null;
    }

    if (!(hpa instanceof HorizontalPodAutoscaler)) {
      logger.error("[HpaDetails]: passed object that is not an instanceof HorizontalPodAutoscaler", hpa);

      return null;
    }

    const { scaleTargetRef } = hpa.spec;

    return (
      <div className="HpaDetails">
        <DrawerItem name="Reference">
          {scaleTargetRef && (
            <Link to={getDetailsUrl(apiManager.lookupApiLink(scaleTargetRef, hpa))}>
              {scaleTargetRef.kind}/{scaleTargetRef.name}
            </Link>
          )}
        </DrawerItem>

        <DrawerItem name="Min Pods">{hpa.getMinPods()}</DrawerItem>

        <DrawerItem name="Max Pods">{hpa.getMaxPods()}</DrawerItem>

        <DrawerItem name="Replicas">{hpa.getReplicas()}</DrawerItem>

        <DrawerItem name="Status" className="status" labelsOnly>
          {hpa.getReadyConditions().map(({ type, tooltip, isReady }) => (
            <Badge key={type} label={type} tooltip={tooltip} className={cssNames({ [type.toLowerCase()]: isReady })} />
          ))}
        </DrawerItem>

        {(hpa.getMetrics().length !== 0 || hpa.spec?.targetCPUUtilizationPercentage) && (
          <>
            <DrawerTitle>Metrics</DrawerTitle>
            <div className="metrics">{this.renderMetrics()}</div>
          </>
        )}
      </div>
    );
  }
}

export const HorizontalPodAutoscalerDetails = withInjectables<Dependencies, HpaDetailsProps>(
  NonInjectedHorizontalPodAutoscalerDetails,
  {
    getProps: (di, props) => ({
      ...props,
      apiManager: di.inject(apiManagerInjectable),
      getDetailsUrl: di.inject(getDetailsUrlInjectable),
      logger: di.inject(loggerInjectionToken),
      getMetrics: di.inject(getHorizontalPodAutoscalerMetrics),
    }),
  },
);
