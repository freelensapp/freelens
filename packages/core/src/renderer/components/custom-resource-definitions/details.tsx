/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import { CustomResourceDefinition } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Link } from "react-router-dom";
import { Badge } from "../badge";
import { DrawerItem, DrawerTitle } from "../drawer";
import { Input } from "../input";
import type { KubeObjectDetailsProps } from "../kube-object-details";
import { MonacoEditor } from "../monaco-editor";
import { Table, TableCell, TableHead, TableRow } from "../table";

export interface CustomResourceDefinitionDetailsProps extends KubeObjectDetailsProps<CustomResourceDefinition> {}

interface Dependencies {
  logger: Logger;
}

@observer
class NonInjectedCustomResourceDefinitionDetails extends React.Component<
  CustomResourceDefinitionDetailsProps & Dependencies
> {
  render() {
    const { object: crd } = this.props;

    if (!crd) {
      return null;
    }

    if (!(crd instanceof CustomResourceDefinition)) {
      this.props.logger.error(
        "[CustomResourceDefinitionDetails]: passed object that is not an instanceof CustomResourceDefinition",
        crd,
      );

      return null;
    }

    const { plural, singular, kind, listKind } = crd.getNames();
    const printerColumns = crd.getPrinterColumns();
    const validation = crd.getValidation();

    return (
      <div className="CustomResourceDefinitionDetails">
        <DrawerItem name="Group">{crd.getGroup()}</DrawerItem>
        <DrawerItem name="Version">{crd.getVersion()}</DrawerItem>
        <DrawerItem name="Stored versions">{crd.getStoredVersions()}</DrawerItem>
        <DrawerItem name="Scope">{crd.getScope()}</DrawerItem>
        <DrawerItem name="Resource">
          <Link to={crd.getResourceUrl()}>{crd.getResourceTitle()}</Link>
        </DrawerItem>
        <DrawerItem name="Conversion" className="flex gaps align-flex-start">
          <Input multiLine theme="round-black" className="box grow" value={crd.getConversion()} readOnly />
        </DrawerItem>
        <DrawerItem name="Conditions" className="conditions" labelsOnly>
          {crd.getConditions().map((condition) => {
            const { type, message, lastTransitionTime, status } = condition;

            return (
              <Badge
                key={type}
                label={type}
                disabled={status === "False"}
                className={type}
                tooltip={
                  <>
                    <p>{message}</p>
                    <p>
                      Last transition time:
                      {lastTransitionTime}
                    </p>
                  </>
                }
              />
            );
          })}
        </DrawerItem>
        <DrawerTitle>Names</DrawerTitle>
        <Table selectable className="names box grow">
          <TableHead>
            <TableCell>plural</TableCell>
            <TableCell>singular</TableCell>
            <TableCell>kind</TableCell>
            <TableCell>listKind</TableCell>
          </TableHead>
          <TableRow>
            <TableCell>{plural}</TableCell>
            <TableCell>{singular}</TableCell>
            <TableCell>{kind}</TableCell>
            <TableCell>{listKind}</TableCell>
          </TableRow>
        </Table>
        {printerColumns.length > 0 && (
          <>
            <DrawerTitle>Additional Printer Columns</DrawerTitle>
            <Table selectable className="printer-columns box grow">
              <TableHead>
                <TableCell className="name">Name</TableCell>
                <TableCell className="type">Type</TableCell>
                <TableCell className="json-path">JSON Path</TableCell>
              </TableHead>
              {printerColumns.map((column, index) => {
                const { name, type, jsonPath } = column;

                return (
                  <TableRow key={index}>
                    <TableCell className="name">{name}</TableCell>
                    <TableCell className="type">{type}</TableCell>
                    <TableCell className="json-path">
                      <Badge label={jsonPath} />
                    </TableCell>
                  </TableRow>
                );
              })}
            </Table>
          </>
        )}
        {validation && (
          <>
            <DrawerTitle>Validation</DrawerTitle>
            <MonacoEditor readOnly value={validation} style={{ height: 400 }} />
          </>
        )}
      </div>
    );
  }
}

export const CustomResourceDefinitionDetails = withInjectables<Dependencies, CustomResourceDefinitionDetailsProps>(
  NonInjectedCustomResourceDefinitionDetails,
  {
    getProps: (di, props) => ({
      ...props,
      logger: di.inject(loggerInjectionToken),
    }),
  },
);
