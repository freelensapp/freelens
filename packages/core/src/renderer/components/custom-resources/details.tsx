/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import "./details.scss";

import type { AdditionalPrinterColumnsV1, KubeObjectMetadata, KubeObjectStatus } from "@freelensapp/kube-object";
import { CustomResourceDefinition, KubeObject } from "@freelensapp/kube-object";
import type { Logger } from "@freelensapp/logger";
import { loggerInjectionToken } from "@freelensapp/logger";
import type { StrictReactNode } from "@freelensapp/utilities";
import { cssNames, safeJSONPathValue } from "@freelensapp/utilities";
import { withInjectables } from "@ogre-tools/injectable-react";
import { observer } from "mobx-react";
import React from "react";
import { Badge } from "../badge";
import { DrawerItem } from "../drawer";
import { Input } from "../input";
import type { KubeObjectDetailsProps } from "../kube-object-details";

export interface CustomResourceDetailsProps extends KubeObjectDetailsProps<KubeObject> {
  crd?: CustomResourceDefinition;
}

function convertSpecValue(value: unknown): StrictReactNode {
  if (Array.isArray(value)) {
    return (
      <ul>
        {value.map((value, index) => (
          <li key={index}>{convertSpecValue(value)}</li>
        ))}
      </ul>
    );
  }

  if (typeof value === "object") {
    return <Input readOnly multiLine theme="round-black" className="box grow" value={JSON.stringify(value, null, 2)} />;
  }

  if (typeof value === "boolean" || typeof value === "string" || typeof value === "number") {
    return value.toString();
  }

  return null;
}

interface Dependencies {
  logger: Logger;
}

@observer
class NonInjectedCustomResourceDetails extends React.Component<CustomResourceDetailsProps & Dependencies> {
  renderAdditionalColumns(resource: KubeObject, columns: AdditionalPrinterColumnsV1[]) {
    return columns.map(({ name, jsonPath }) => (
      <DrawerItem key={name} name={name}>
        {convertSpecValue(safeJSONPathValue(resource, jsonPath))}
      </DrawerItem>
    ));
  }

  renderStatus(cr: KubeObject, columns: AdditionalPrinterColumnsV1[]) {
    const customResource = cr as KubeObject<KubeObjectMetadata, KubeObjectStatus, unknown>;
    const showStatus =
      !columns.find((column) => column.name == "Status") && Array.isArray(customResource.status?.conditions);

    if (!showStatus) {
      return null;
    }

    const conditions = customResource.status?.conditions
      ?.filter(({ type, reason }) => type || reason)
      .map(({ type, reason, message, status }) => ({
        kind: type || reason || "<unknown>",
        message,
        status,
      }))
      .map(({ kind, message, status }, index) => (
        <Badge
          key={`${kind}${index}`}
          label={kind}
          disabled={status === "False"}
          className={kind.toLowerCase()}
          tooltip={message}
        />
      ));

    return (
      <DrawerItem name="Status" className="status" labelsOnly>
        {conditions}
      </DrawerItem>
    );
  }

  render() {
    const {
      props: { object, crd, logger },
    } = this;

    if (!object || !crd) {
      return null;
    }

    if (!(object instanceof KubeObject)) {
      logger.error("[CrdResourceDetails]: passed object that is not an instanceof KubeObject", object);

      return null;
    }

    if (!(crd instanceof CustomResourceDefinition)) {
      logger.error("[CrdResourceDetails]: passed crd that is not an instanceof CustomResourceDefinition", crd);

      return null;
    }

    const extraColumns = crd.getPrinterColumns();

    return (
      <div className={cssNames("CustomResourceDetails", crd.getResourceKind())}>
        {this.renderAdditionalColumns(object, extraColumns)}
        {this.renderStatus(object, extraColumns)}
      </div>
    );
  }
}

export const CustomResourceDetails = withInjectables<Dependencies, CustomResourceDetailsProps>(
  NonInjectedCustomResourceDetails,
  {
    getProps: (di, props) => ({
      ...props,
      logger: di.inject(loggerInjectionToken),
    }),
  },
);
