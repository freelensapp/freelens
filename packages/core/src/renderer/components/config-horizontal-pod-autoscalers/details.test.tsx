/**
 * Copyright (c) Freelens Authors. All rights reserved.
 * Copyright (c) OpenLens Authors. All rights reserved.
 * Licensed under MIT License. See LICENSE in root directory for more information.
 */

import { HorizontalPodAutoscaler } from "@freelensapp/kube-object";
import type { StrictReactNode } from "@freelensapp/utilities";
import type { RenderResult } from "@testing-library/react";
import React from "react";
import directoryForKubeConfigsInjectable from "../../../common/app-paths/directory-for-kube-configs/directory-for-kube-configs.injectable";
import directoryForUserDataInjectable from "../../../common/app-paths/directory-for-user-data/directory-for-user-data.injectable";
import { Cluster } from "../../../common/cluster/cluster";
import hostedClusterInjectable from "../../cluster-frame-context/hosted-cluster.injectable";
import { getDiForUnitTesting } from "../../getDiForUnitTesting";
import storesAndApisCanBeCreatedInjectable from "../../stores-apis-can-be-created.injectable";
import type { DiRender } from "../test-utils/renderFor";
import { renderFor } from "../test-utils/renderFor";
import { HorizontalPodAutoscalerDetails } from "./details";

jest.mock("react-router-dom", () => ({
  Link: ({ children }: { children: StrictReactNode }) => children,
}));

const hpaV2 = {
  apiVersion: "autoscaling/v2",
  kind: "HorizontalPodAutoscaler",
  metadata: {
    name: "hpav2",
    resourceVersion: "1",
    uid: "hpav2",
    namespace: "default",
    selfLink: "/apis/autoscaling/v2/namespaces/default/horizontalpodautoscalers/hpav2",
  },
  spec: {
    maxReplicas: 10,
    scaleTargetRef: {
      kind: "Deployment",
      name: "hpav2deployment",
      apiVersion: "apps/v1",
    },
  },
};

describe("<HpaDetails/>", () => {
  let result: RenderResult;
  let render: DiRender;

  beforeEach(() => {
    const di = getDiForUnitTesting();

    di.override(directoryForUserDataInjectable, () => "/some-user-store-path");
    di.override(directoryForKubeConfigsInjectable, () => "/some-kube-configs");
    di.override(storesAndApisCanBeCreatedInjectable, () => true);
    di.override(
      hostedClusterInjectable,
      () =>
        new Cluster({
          contextName: "some-context-name",
          id: "some-cluster-id",
          kubeConfigPath: "/some-path-to-a-kubeconfig",
        }),
    );

    render = renderFor(di);
  });

  it("renders", () => {
    const hpa = new HorizontalPodAutoscaler(hpaV2);

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.baseElement).toMatchSnapshot();
  });

  it("does not show metrics table if no metrics found", () => {
    const hpa = new HorizontalPodAutoscaler(hpaV2);

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.queryByTestId("hpa-metrics")).toBeNull();
  });

  it("shows proper metric name for autoscaling/v1", () => {
    const hpa = new HorizontalPodAutoscaler({
      apiVersion: "autoscaling/v1",
      kind: "HorizontalPodAutoscaler",
      metadata: {
        name: "hpav1",
        resourceVersion: "1",
        uid: "hpav1",
        namespace: "default",
        selfLink: "/apis/autoscaling/v1/namespaces/default/horizontalpodautoscalers/hpav1",
      },
      spec: {
        maxReplicas: 10,
        scaleTargetRef: {
          kind: "Deployment",
          name: "hpav1deployment",
          apiVersion: "apps/v1",
        },
        targetCPUUtilizationPercentage: 80,
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("CPU Utilization percentage")).toBeInTheDocument();
  });

  it("shows proper metric name for container resource metrics", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "ContainerResource",
            containerResource: {
              name: "cpu",
              container: "nginx",
              target: {
                type: "Utilization",
                averageUtilization: 60,
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("Resource cpu on Pods")).toBeInTheDocument();
  });

  it("shows proper metric name for resource metrics", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "Resource",
            resource: {
              name: "cpu",
              target: {
                type: "Utilization",
                averageUtilization: 50,
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("Resource cpu on Pods")).toBeInTheDocument();
  });

  it("shows proper metric name for pod metrics for hpa v2", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "Pods",
            pods: {
              metric: {
                name: "packets-per-second",
              },
              target: {
                type: "AverageValue",
                averageValue: "1k",
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("packets-per-second on Pods")).toBeInTheDocument();
  });

  it("shows proper metric name for pod metrics for hpa v2beta1", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "Pods",
            pods: {
              metricName: "packets-per-second",
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("packets-per-second on Pods")).toBeInTheDocument();
  });

  it("shows proper metric name for object metrics for hpa v2", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "Object",
            object: {
              metric: {
                name: "requests-per-second",
              },
              target: {
                type: "Value",
                value: "10k",
              },
              describedObject: {
                kind: "Service",
                name: "nginx",
                apiVersion: "v1",
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText(/requests-per-second/)).toHaveTextContent("requests-per-second onService/nginx");
  });

  it("shows proper metric name for object metrics for hpa v2beta1", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "Object",
            object: {
              metricName: "requests-per-second",
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText("requests-per-second")).toBeInTheDocument();
  });

  it("shows proper metric name for external metrics for hpa v2", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "External",
            external: {
              metric: {
                name: "queue_messages_ready",
                selector: {
                  matchLabels: { queue: "worker_tasks" },
                },
              },
              target: {
                type: "AverageValue",
                averageValue: "30",
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText('queue_messages_ready on {"matchLabels":{"queue":"worker_tasks"}}')).toBeInTheDocument();
  });

  it("shows proper metric name for external metrics for hpa v2beta1", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            type: "External",
            external: {
              metricName: "queue_messages_ready",
              metricSelector: {
                matchLabels: { queue: "worker_tasks" },
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.getByText('queue_messages_ready on {"matchLabels":{"queue":"worker_tasks"}}')).toBeInTheDocument();
  });

  it("shows unknown metrics with lack of metric type", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          // @ts-ignore
          {
            resource: {
              name: "cpu",
              target: {
                type: "Utilization",
                averageUtilization: 50,
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.baseElement).toMatchSnapshot();
  });

  it("shows unknown metrics with with unusual type", () => {
    const hpa = new HorizontalPodAutoscaler({
      ...hpaV2,
      spec: {
        ...hpaV2.spec,
        metrics: [
          {
            // @ts-ignore
            type: "Unusual",
            resource: {
              name: "cpu",
              target: {
                type: "Utilization",
                averageUtilization: 50,
              },
            },
          },
        ],
      },
    });

    result = render(<HorizontalPodAutoscalerDetails object={hpa} />);

    expect(result.baseElement).toMatchSnapshot();
  });
});
