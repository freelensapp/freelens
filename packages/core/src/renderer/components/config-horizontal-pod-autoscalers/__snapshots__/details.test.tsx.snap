// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<HpaDetails/> renders 1`] = `
<body>
  <div>
    <div
      class="HpaDetails"
    >
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Reference
        </span>
        <span
          class="value"
        >
          Deployment
          /
          hpav2deployment
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Min Pods
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Max Pods
        </span>
        <span
          class="value"
        >
          10
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Replicas
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
    </div>
  </div>
</body>
`;

exports[`<HpaDetails/> shows unknown metrics with lack of metric type 1`] = `
<body>
  <div>
    <div
      class="HpaDetails"
    >
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Reference
        </span>
        <span
          class="value"
        >
          Deployment
          /
          hpav2deployment
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Min Pods
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Max Pods
        </span>
        <span
          class="value"
        >
          10
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Replicas
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
      <div
        class="DrawerTitle title"
      >
        Metrics
      </div>
      <div
        class="metrics"
      >
        <div
          class="Table flex column scrollable autoSize"
        >
          <div
            class="TableHead sticky flat"
          >
            <div
              class="TableCell name"
            >
              Name
            </div>
            <div
              class="TableCell metrics"
            >
              Current / Target
            </div>
          </div>
          <div
            class="TableRow"
          >
            <div
              class="TableCell name"
            >
              unknown
            </div>
            <div
              class="TableCell metrics"
            >
              unknown / unknown
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;

exports[`<HpaDetails/> shows unknown metrics with with unusual type 1`] = `
<body>
  <div>
    <div
      class="HpaDetails"
    >
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Reference
        </span>
        <span
          class="value"
        >
          Deployment
          /
          hpav2deployment
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Min Pods
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Max Pods
        </span>
        <span
          class="value"
        >
          10
        </span>
      </div>
      <div
        class="DrawerItem"
      >
        <span
          class="name"
        >
          Replicas
        </span>
        <span
          class="value"
        >
          0
        </span>
      </div>
      <div
        class="DrawerTitle title"
      >
        Metrics
      </div>
      <div
        class="metrics"
      >
        <div
          class="Table flex column scrollable autoSize"
        >
          <div
            class="TableHead sticky flat"
          >
            <div
              class="TableCell name"
            >
              Name
            </div>
            <div
              class="TableCell metrics"
            >
              Current / Target
            </div>
          </div>
          <div
            class="TableRow"
          >
            <div
              class="TableCell name"
            >
              unknown
            </div>
            <div
              class="TableCell metrics"
            >
              unknown / unknown
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
